### PR TITLE
[jenkins] Implement support for slack notifications.

### PR DESCRIPTION
**Pull request by rolfbjarne**
Implement support for slack notifications.

Notifications will only be sent when something fails, and unfortunately I wasn't able to make notifications ping due to https://github.com/jenkinsci/slack-plugin/issues/353.

### CODE REVIEW

**Comment by xamarin-release-manager** [comment]
✅ [Jenkins job](http://xamarin-jenkins.guest.corp.microsoft.com:8080/job/macios/job/PR-4369/1/display/redirect) (on internal Jenkins) succeeded

✅ [Build succeeded](http://xamarin-jenkins.guest.corp.microsoft.com:8080/job/macios/job/PR-4369/1//console)
✅ [API Diff (from stable)](http://xamarin-storage/jenkins/xamarin-macios/PR-4369/06e8c4de7f79c9f8641851dd1f7095240965411b/1/jenkins-results/api-diff/index.html)
✅ [API Diff (from PR only)](http://xamarin-storage/jenkins/xamarin-macios/PR-4369/06e8c4de7f79c9f8641851dd1f7095240965411b/1/jenkins-results/apicomparison/index.html) (please review changes)
✅ [Generator Diff](http://xamarin-storage/jenkins/xamarin-macios/PR-4369/06e8c4de7f79c9f8641851dd1f7095240965411b/1/jenkins-results/generator-diff/index.html) (only version changes)
ℹ️ Test run skipped: Not running tests here because they're run on public Jenkins.

**Comment by spouliot** [comment]
That's waiting on #4278 right ?

**Comment by rolfbjarne** [comment]
@spouliot not really, I was just looking into how to best format the message. The most annoying thing about Jenkins is that the only way to test things is to create pull requests, I can't do it locally 😒

**Comment by xamarin-release-manager** [comment]
🔥 [Jenkins job](http://xamarin-jenkins.guest.corp.microsoft.com:8080/job/macios/job/PR-4369/3/display/redirect) (on internal Jenkins) failed in stage 'Package XM tests' 🔥 : hudson.AbortException: script returned exit code 2

✅ [Build succeeded](http://xamarin-jenkins.guest.corp.microsoft.com:8080/job/macios/job/PR-4369/3//console)
✅ [API Diff (from stable)](http://xamarin-storage/jenkins/xamarin-macios/PR-4369/fb0acca53e69366a8a4def61a55b2ef8f2678b4c/3/jenkins-results/api-diff/index.html)
✅ [API Diff (from PR only)](http://xamarin-storage/jenkins/xamarin-macios/PR-4369/fb0acca53e69366a8a4def61a55b2ef8f2678b4c/3/jenkins-results/apicomparison/index.html) (please review changes)
✅ [Generator Diff](http://xamarin-storage/jenkins/xamarin-macios/PR-4369/fb0acca53e69366a8a4def61a55b2ef8f2678b4c/3/jenkins-results/generator-diff/index.html) (only version changes)

**Comment by xamarin-release-manager** [comment]
🔥 [Jenkins job](http://xamarin-jenkins.guest.corp.microsoft.com:8080/job/macios/job/PR-4369/4/display/redirect) (on internal Jenkins) failed in stage 'Package XM tests' 🔥 : hudson.AbortException: script returned exit code 2

✅ [Build succeeded](http://xamarin-jenkins.guest.corp.microsoft.com:8080/job/macios/job/PR-4369/4//console)
✅ [API Diff (from stable)](http://xamarin-storage/jenkins/xamarin-macios/PR-4369/2ecee8533984dd59fa6669fc35fbb098890146d7/4/jenkins-results/api-diff/index.html)
✅ [API Diff (from PR only)](http://xamarin-storage/jenkins/xamarin-macios/PR-4369/2ecee8533984dd59fa6669fc35fbb098890146d7/4/jenkins-results/apicomparison/index.html) (please review changes)
✅ [Generator Diff](http://xamarin-storage/jenkins/xamarin-macios/PR-4369/2ecee8533984dd59fa6669fc35fbb098890146d7/4/jenkins-results/generator-diff/index.html) (only version changes)

**Comment by xamarin-release-manager** [comment]
🔥 [Jenkins job](http://xamarin-jenkins.guest.corp.microsoft.com:8080/job/macios/job/PR-4369/6/display/redirect) (on internal Jenkins) failed in stage 'Package XM tests' 🔥 : hudson.AbortException: script returned exit code 2

✅ [Build succeeded](http://xamarin-jenkins.guest.corp.microsoft.com:8080/job/macios/job/PR-4369/6//console)
✅ [API Diff (from stable)](http://xamarin-storage/jenkins/xamarin-macios/PR-4369/269ce35011b6fc99d0d7b2bbe01722d52198df98/6/jenkins-results/api-diff/index.html)
✅ [API Diff (from PR only)](http://xamarin-storage/jenkins/xamarin-macios/PR-4369/269ce35011b6fc99d0d7b2bbe01722d52198df98/6/jenkins-results/apicomparison/index.html) (please review changes)
✅ [Generator Diff](http://xamarin-storage/jenkins/xamarin-macios/PR-4369/269ce35011b6fc99d0d7b2bbe01722d52198df98/6/jenkins-results/generator-diff/index.html) (only version changes)

**Comment by xamarin-release-manager** [comment]
🔥 [Jenkins job](http://xamarin-jenkins.guest.corp.microsoft.com:8080/job/macios/job/PR-4369/7/display/redirect) (on internal Jenkins) failed in stage 'Package XM tests' 🔥 : hudson.AbortException: script returned exit code 2

✅ [Build succeeded](http://xamarin-jenkins.guest.corp.microsoft.com:8080/job/macios/job/PR-4369/7//console)
✅ [API Diff (from stable)](http://xamarin-storage/jenkins/xamarin-macios/PR-4369/133fafcadcdcda867e56f7f4ee7a45685d7facbc/7/jenkins-results/api-diff/index.html)
✅ [API Diff (from PR only)](http://xamarin-storage/jenkins/xamarin-macios/PR-4369/133fafcadcdcda867e56f7f4ee7a45685d7facbc/7/jenkins-results/apicomparison/index.html) (please review changes)
✅ [Generator Diff](http://xamarin-storage/jenkins/xamarin-macios/PR-4369/133fafcadcdcda867e56f7f4ee7a45685d7facbc/7/jenkins-results/generator-diff/index.html) (only version changes)

**Comment by xamarin-release-manager** [comment]
🔥 [Jenkins job](http://xamarin-jenkins.guest.corp.microsoft.com:8080/job/macios/job/PR-4369/8/display/redirect) (on internal Jenkins) failed in stage 'Package XM tests' 🔥 : hudson.AbortException: script returned exit code 2

✅ [Build succeeded](http://xamarin-jenkins.guest.corp.microsoft.com:8080/job/macios/job/PR-4369/8//console)
✅ [API Diff (from stable)](http://xamarin-storage/jenkins/xamarin-macios/PR-4369/68a606506826a79094d78a5ab6e5f95305163ec9/8/jenkins-results/api-diff/index.html)
✅ [API Diff (from PR only)](http://xamarin-storage/jenkins/xamarin-macios/PR-4369/68a606506826a79094d78a5ab6e5f95305163ec9/8/jenkins-results/apicomparison/index.html) (please review changes)
✅ [Generator Diff](http://xamarin-storage/jenkins/xamarin-macios/PR-4369/68a606506826a79094d78a5ab6e5f95305163ec9/8/jenkins-results/generator-diff/index.html) (only version changes)

**Comment by xamarin-release-manager** [comment]
🔥 [Jenkins job](http://xamarin-jenkins.guest.corp.microsoft.com:8080/job/macios/job/PR-4369/9/display/redirect) (on internal Jenkins) failed in stage 'Package XM tests' 🔥 : hudson.AbortException: script returned exit code 2

✅ [Build succeeded](http://xamarin-jenkins.guest.corp.microsoft.com:8080/job/macios/job/PR-4369/9//console)
✅ [API Diff (from stable)](http://xamarin-storage/jenkins/xamarin-macios/PR-4369/02a1910df8a89250a3a1b8ffaea7412a6c97d585/9/jenkins-results/api-diff/index.html)
✅ [API Diff (from PR only)](http://xamarin-storage/jenkins/xamarin-macios/PR-4369/02a1910df8a89250a3a1b8ffaea7412a6c97d585/9/jenkins-results/apicomparison/index.html) (please review changes)
✅ [Generator Diff](http://xamarin-storage/jenkins/xamarin-macios/PR-4369/02a1910df8a89250a3a1b8ffaea7412a6c97d585/9/jenkins-results/generator-diff/index.html) (only version changes)

**Comment by xamarin-release-manager** [comment]
🔥 [Jenkins job](http://xamarin-jenkins.guest.corp.microsoft.com:8080/job/macios/job/PR-4369/10/display/redirect) (on internal Jenkins) failed in stage 'Package XM tests' 🔥 : hudson.AbortException: script returned exit code 2

✅ [Build succeeded](http://xamarin-jenkins.guest.corp.microsoft.com:8080/job/macios/job/PR-4369/10//console)
✅ [API Diff (from stable)](http://xamarin-storage/jenkins/xamarin-macios/PR-4369/02a1910df8a89250a3a1b8ffaea7412a6c97d585/10/jenkins-results/api-diff/index.html)
✅ [API Diff (from PR only)](http://xamarin-storage/jenkins/xamarin-macios/PR-4369/02a1910df8a89250a3a1b8ffaea7412a6c97d585/10/jenkins-results/apicomparison/index.html) (please review changes)
✅ [Generator Diff](http://xamarin-storage/jenkins/xamarin-macios/PR-4369/02a1910df8a89250a3a1b8ffaea7412a6c97d585/10/jenkins-results/generator-diff/index.html) (only version changes)

**Comment by xamarin-release-manager** [comment]
🔥 [Jenkins job](http://xamarin-jenkins.guest.corp.microsoft.com:8080/job/macios/job/PR-4369/11/display/redirect) (on internal Jenkins) failed in stage 'Package XM tests' 🔥 : hudson.AbortException: script returned exit code 2

✅ [Build succeeded](http://xamarin-jenkins.guest.corp.microsoft.com:8080/job/macios/job/PR-4369/11//console)
✅ [API Diff (from stable)](http://xamarin-storage/jenkins/xamarin-macios/PR-4369/12e1e3eaf06554d62cde54110a03b33129b253e2/11/jenkins-results/api-diff/index.html)
✅ [API Diff (from PR only)](http://xamarin-storage/jenkins/xamarin-macios/PR-4369/12e1e3eaf06554d62cde54110a03b33129b253e2/11/jenkins-results/apicomparison/index.html) (please review changes)
✅ [Generator Diff](http://xamarin-storage/jenkins/xamarin-macios/PR-4369/12e1e3eaf06554d62cde54110a03b33129b253e2/11/jenkins-results/generator-diff/index.html) (only version changes)

**Comment by xamarin-release-manager** [comment]
🔥 [Jenkins job](http://xamarin-jenkins.guest.corp.microsoft.com:8080/job/macios/job/PR-4369/12/display/redirect) (on internal Jenkins) failed in stage 'Package XM tests' 🔥 : hudson.AbortException: script returned exit code 2

✅ [Build succeeded](http://xamarin-jenkins.guest.corp.microsoft.com:8080/job/macios/job/PR-4369/12//console)
✅ [API Diff (from stable)](http://xamarin-storage/jenkins/xamarin-macios/PR-4369/64a18be37ff03a6c957bc14c5448592b83ed24ca/12/jenkins-results/api-diff/index.html)
✅ [API Diff (from PR only)](http://xamarin-storage/jenkins/xamarin-macios/PR-4369/64a18be37ff03a6c957bc14c5448592b83ed24ca/12/jenkins-results/apicomparison/index.html) (please review changes)
✅ [Generator Diff](http://xamarin-storage/jenkins/xamarin-macios/PR-4369/64a18be37ff03a6c957bc14c5448592b83ed24ca/12/jenkins-results/generator-diff/index.html) (only version changes)

**Comment by xamarin-release-manager** [comment]
🔥 [Jenkins job](http://xamarin-jenkins.guest.corp.microsoft.com:8080/job/macios/job/PR-4369/13/display/redirect) (on internal Jenkins) failed in stage 'Package XM tests' 🔥 : hudson.AbortException: script returned exit code 2

✅ [Build succeeded](http://xamarin-jenkins.guest.corp.microsoft.com:8080/job/macios/job/PR-4369/13//console)
✅ [API Diff (from stable)](http://xamarin-storage/jenkins/xamarin-macios/PR-4369/b1b854d6cdd39f8c83450435ddc98d66a85a7157/13/jenkins-results/api-diff/index.html)
✅ [API Diff (from PR only)](http://xamarin-storage/jenkins/xamarin-macios/PR-4369/b1b854d6cdd39f8c83450435ddc98d66a85a7157/13/jenkins-results/apicomparison/index.html) (please review changes)
✅ [Generator Diff](http://xamarin-storage/jenkins/xamarin-macios/PR-4369/b1b854d6cdd39f8c83450435ddc98d66a85a7157/13/jenkins-results/generator-diff/index.html) (only version changes)

**Comment by xamarin-release-manager** [comment]
🔥 [Jenkins job](http://xamarin-jenkins.guest.corp.microsoft.com:8080/job/macios/job/PR-4369/14/display/redirect) (on internal Jenkins) failed in stage 'Package XM tests' 🔥 : hudson.AbortException: script returned exit code 2

✅ [Build succeeded](http://xamarin-jenkins.guest.corp.microsoft.com:8080/job/macios/job/PR-4369/14//console)
✅ [API Diff (from stable)](http://xamarin-storage/jenkins/xamarin-macios/PR-4369/3b8bd00a48f65f25d29e2d9efe48965a14d57909/14/jenkins-results/api-diff/index.html)
✅ [API Diff (from PR only)](http://xamarin-storage/jenkins/xamarin-macios/PR-4369/3b8bd00a48f65f25d29e2d9efe48965a14d57909/14/jenkins-results/apicomparison/index.html) (please review changes)
✅ [Generator Diff](http://xamarin-storage/jenkins/xamarin-macios/PR-4369/3b8bd00a48f65f25d29e2d9efe48965a14d57909/14/jenkins-results/generator-diff/index.html) (only version changes)

**Comment by xamarin-release-manager** [comment]
🔥 [Jenkins job](http://xamarin-jenkins.guest.corp.microsoft.com:8080/job/macios/job/PR-4369/15/display/redirect) (on internal Jenkins) failed in stage 'Package XM tests' 🔥 : hudson.AbortException: script returned exit code 2

✅ [Build succeeded](http://xamarin-jenkins.guest.corp.microsoft.com:8080/job/macios/job/PR-4369/15//console)
✅ [API Diff (from stable)](http://xamarin-storage/jenkins/xamarin-macios/PR-4369/5ffb588838eb915d084ee341ec8ceaf466c705d4/15/jenkins-results/api-diff/index.html)
✅ [API Diff (from PR only)](http://xamarin-storage/jenkins/xamarin-macios/PR-4369/5ffb588838eb915d084ee341ec8ceaf466c705d4/15/jenkins-results/apicomparison/index.html) (please review changes)
✅ [Generator Diff](http://xamarin-storage/jenkins/xamarin-macios/PR-4369/5ffb588838eb915d084ee341ec8ceaf466c705d4/15/jenkins-results/generator-diff/index.html) (only version changes)

**Comment by rolfbjarne** [comment]
Rendered in Slack:

screenshot 2018-08-02 15 49 16

**Comment by rolfbjarne** [comment]
This is ready for review now (a rendered notification is shown in my previous comment).

File: jenkins/Jenkinsfile [hunk 1]
```diff
@@ -106,6 +106,63 @@ def appendFileComment (comment)
     writeFile (file: commentFile, text: comment)
 }
 
+def markdownToSlack (value)
+{
+    def tmpfile = "/tmp/slacker.md"
+    writeFile (file: tmpfile, text: value)
+    try {
+        // this monstruousity converts markdown's [text](url) to slack's <url|text>
+        sh ("sed -i '' 's/[[]\\(.*\\)[]][\\(]\\(.*\\)[\\)]/<\\2|\\1>/' '${tmpfile}'")
```
Comment (spouliot) [inline jenkins/Jenkinsfile]: 👹

**Comment by xamarin-release-manager** [comment]
✅ [Jenkins job](http://xamarin-jenkins.guest.corp.microsoft.com:8080/job/macios/job/PR-4369/16/display/redirect) (on internal Jenkins) succeeded

✅ [Build succeeded](http://xamarin-jenkins.guest.corp.microsoft.com:8080/job/macios/job/PR-4369/16//console)
✅ [API Diff (from stable)](http://xamarin-storage/jenkins/xamarin-macios/PR-4369/96bc3445feb428ec02017f0b0c708bf8e75871fd/16/jenkins-results/api-diff/index.html)
✅ [API Diff (from PR only)](http://xamarin-storage/jenkins/xamarin-macios/PR-4369/96bc3445feb428ec02017f0b0c708bf8e75871fd/16/jenkins-results/apicomparison/index.html) (please review changes)
✅ [Generator Diff](http://xamarin-storage/jenkins/xamarin-macios/PR-4369/96bc3445feb428ec02017f0b0c708bf8e75871fd/16/jenkins-results/generator-diff/index.html) (only version changes)
✅ [Test run succeeded](http://xamarin-storage/jenkins/xamarin-macios/PR-4369/96bc3445feb428ec02017f0b0c708bf8e75871fd/16/jenkins-results/tests/index.html)

**Pull request merged**